File: Problem_1 Find Town_Judge.py
# // Time Complexity : O(V+E) where V and E are vertices and Edges of the graph
# // Space Complexity : O(V)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        indegree = [0 for i in range(n)]
        #Iterate over all the Elements of Trust Array
        for t in trust:
            indegree[t[1]-1]+=1
            indegree[t[0]-1]-=1
        for i in range(len(indegree)):
            #if the edge has all incoming nodes
            if indegree[i] == n-1:
                return i+1
        # if no edge has all incoming nodes, return -1
        return -1

File: Problem1_FindTownJudge.java
/*
Time complexity: O(N) ; we need to traverse the given trust [N = m+n]
Space Complexity: O(N) ; we use array of length N
*/
class Solution {
    public int findJudge(int n, int[][] trust) {
        
        //Base
        if(trust.length==0 && n==1){
            return 1;
        }
        int m = trust.length;
        int[] indegrees = new int[n+1];
        int[] outdegrees = new int[n+1];
        // System.out.println("Length of both arrays: "+indegrees.length);
        
        // List<Integer> indegrees = new List<>(m); //Else only Degrees
        for(int[] t : trust ){
            // for(int j = 0; j<n ; j++){
            // System.out.print(t[0] + " trusts "+ t[1]);
            // outdegrees[t[0]]++;
            indegrees[t[1]]++;
            // System.out.println(indegrees+ " " + outdegrees);
            // }
        }
        
        for(int j = 0; j<indegrees.length ; j++){
            indegrees[j] = indegrees[j] - outdegrees[j];
        }
        for(int j = 0; j<indegrees.length ; j++){
            if(indegrees[j] == n-1) return j;
        }
        return -1;
    }
}
